content/posts/2022-08-02-paxos/index.md

```diff
@@ -0,0 +1,134 @@
+
+It turns out that there is a solution even in the case of an unreliable network, and that is the so-called `Single-decree Paxos` or `Synod,` which is a base block for building more complicated protocols. Synod protocol itself is actually quite simple; it has two phases (usually referred to as `Phase 1` and `Phase 2`) and three types of messages (`prepare,` `accept,` `decide` with positive and negative responses for the first two), and it can be described in one screen of pseudocode: <http://nil.csail.mit.edu/6.824/2015/notes/paxos-code.html>. It also turns out that `Synod` may never terminate, being forever bouncing between phases, but in practice, it is not a problem, and with some randomization, it quickly converges.
+
+**Atomic broadcast**. Following the Wikipedia definition, atomic broadcast is a broadcast in which all correct processes in a system of multiple processes receive the same set of messages in the same order. Or, in less strict terms, it is a fault-tolerant replication. And that is precisely the problem that Raft, Paxos, and ZAB are solving, so the problem's scope is bigger than a single-value consensus. The atomic broadcast protocol also needs to deal with the detection of dead nodes, node recoveries after the restart, split-brain protection, etc.
+
+**Paxos**. Paxos, or more precisely, multi-Paxos, is one of the first protocols to solve atomic broadcast and probably the most famous one. The protocol defines several types of nodes, namely:
+
+- Proposers – replication "source" that executes the client request and coordinates its further execution until the client's value is fully replicated. Paxos also selects a distinguished Proposer who is referred to as a Leader.
+- Acceptors – replication "destination" for fresh and not necessarily committed values
+- Learners – replication "destination" for definitely committed values
+
+In a practical system, some or all of the roles are usually colocated in the same node. The protocol gives much freedom about exact implementation, but that freedom may lead to hard-to-reason bugs.
+
+**Raft**. Raft solves the same algorithmic problem but in a more engineering-friendly way. Instead of talking about all possible roles like Paxos, Raft is described like a usual replication setup: there is a primary (called leader) and a set of replicas (called followers). When followers do not hear from the leader for a while, one of them becomes a candidate and starts voting to become a leader. Each election procedure is stamped with a number called `term.` But, underneath, it is a similar concept to Paxos, just put together in a well-thought-out manner:
+each node collocates Proposer/Acceptor/Learner;
+voting is the same `Phase 1` from Paxos (albeit with different term selection rules);
```

Comment:
Well, Paxos allows any node to be elected (which is said below), so saying it is the same is not accurate.

The main thing in Paxos-vs-Raft talk IMV is this (allowing any node to be elected) + as a consequence, necessity for the leader to learn the latest accepted value for each not yet committed cell instead of just forcing its own log on other nodes. Not sure how to put that shortly. Apparently this is meant under "some recovery problems are mitigated...".